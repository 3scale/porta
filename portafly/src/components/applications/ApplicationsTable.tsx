import React from 'react'
import { Table, TableHeader, TableBody } from '@patternfly/react-table'
import { SimpleEmptyState } from 'components'
import { IApplication } from 'types'

export interface IApplicationsTable {
  applications: IApplication[]
}

const ApplicationsTable: React.FunctionComponent<IApplicationsTable> = ({ applications }) => {
  if (applications.length === 0) {
    return <SimpleEmptyState msg="There are no Applications" />
  }

  const COLUMNS = ['Name', 'State', 'Account', 'Plan', 'Created at']

  const rows = applications.map(app => [app.name, app.state, app.account, app.plan.name, app.created_at])

  return (
    <Table aria-label="Applications list" cells={COLUMNS} rows={rows}>
      <TableHeader />
      <TableBody />
    </Table>
  )
}

export { ApplicationsTable }
