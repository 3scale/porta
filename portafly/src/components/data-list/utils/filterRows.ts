import { DataListRow, DataListCol, Filters } from 'types/data-list'

export function filterRows(
  rows: DataListRow[],
  filters: Filters,
  columns: DataListCol[]
) {
  let newRows = [...rows]

  Object.keys(filters).forEach((categoryName) => {
    if (newRows.length === 0 || filters[categoryName].length === 0) {
      return
    }

    const colIdx = columns.findIndex((c) => c.categoryName === categoryName)

    if (colIdx === -1) {
      console.warn('Category "%s" doesn\'t correspond to any column in your DataList', categoryName)
      return
    }

    newRows = newRows.filter((r) => {
      const value = (r.cells)[colIdx].toLowerCase()
      return filters[categoryName].some((f) => value.indexOf(f.toLowerCase()) > -1)
    })
  })

  return newRows
}
