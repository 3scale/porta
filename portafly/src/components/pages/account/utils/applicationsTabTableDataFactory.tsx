import React from 'react'
import { sortable } from '@patternfly/react-table'
import { DataListRowGenerator, DataListColumnGenerator, IAccountApplication } from 'types'
import { TFunction } from 'i18next'
import {
  ApplicationPageLink,
  PlanOverviewLink,
  StateLabel,
  ProductLink
} from 'components'

const generateRows: DataListRowGenerator = (applications: IAccountApplication[]) => {
  // Rows and Columns must have the same order
  const mapAccountToRowCell = (application: IAccountApplication) => [
    {
      stringValue: application.name,
      title: <ApplicationPageLink application={application} />
    },
    {
      stringValue: application.product.name,
      title: <ProductLink product={application.product} />
    },
    {
      stringValue: application.plan.name,
      title: <PlanOverviewLink plan={application.plan} />
    },
    application.createdOn,
    {
      stringValue: application.state,
      title: <StateLabel state={application.state} />
    }
  ]

  return applications.map((a) => ({
    id: a.id,
    cells: mapAccountToRowCell(a),
    selected: false
  }))
}

// Filterable columns must have an id equal to its category name
const generateColumns: DataListColumnGenerator = (t: TFunction) => [
  {
    categoryName: 'name',
    title: t('applications.applications_table.name_header'),
    transforms: [sortable]
  },
  {
    categoryName: 'product',
    title: t('applications.applications_table.product_header'),
    transforms: [sortable]
  },
  {
    categoryName: 'plan',
    title: t('applications.applications_table.plan_header'),
    transforms: [sortable]
  },
  {
    categoryName: 'createdOn',
    title: t('applications.applications_table.created_header'),
    transforms: [sortable]
  },
  {
    categoryName: 'state',
    title: t('applications.applications_table.state_header'),
    transforms: [sortable]
  }
]

export { generateColumns, generateRows }
