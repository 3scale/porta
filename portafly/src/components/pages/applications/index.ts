export { default as ApplicationsIndexPage } from './ApplicationsIndexPage'
export * from './ApplicationsDataListTable'
export * from './ApplicationsTable'
export * from './CreateApplicationButton'
export * from './ApplicationPageLink'
