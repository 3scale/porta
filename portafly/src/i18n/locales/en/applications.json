{
  "page_title": "Applications | Portafly",
  "page_title_desc": "Page title of the Applications page",
  "body_title": "Applications",
  "body_title_desc": "Body title of the Applications page",
  "subtitle": "This is the Applications page",
  "subtitle_desc": "Subtitle of the Applications page"
}
