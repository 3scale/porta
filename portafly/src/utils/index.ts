export * from './http'
export * from './reducers'
