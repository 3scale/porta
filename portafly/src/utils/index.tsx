export type Action = {
  type: string,
  payload?: any
}
export type State = Record<string, any>
export type ActionHandlers = Record<string, (state: State, action: Action) => State>

const createReducer = (handlers: ActionHandlers) => (
  (state: Record<string, any>, action: Action): State => (
    handlers.hasOwnProperty(action.type) ? handlers[action.type](state, action) : state
  )
)
const combineReducers = (reducerSlices: Record<string, any>) => (
  (prevState: State, action: Action) => Object.keys(reducerSlices).reduce(
    (nextState, nextProp) => ({
      ...nextState,
      [nextProp]: reducerSlices[nextProp](prevState[nextProp], action)
    }),
    prevState
  ))

export { createReducer, combineReducers }
