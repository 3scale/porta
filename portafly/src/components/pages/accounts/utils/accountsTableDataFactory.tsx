import React from 'react'
import {
  sortable,
  cellWidth
} from '@patternfly/react-table'
import { IDeveloperAccount, DataListRowGenerator, DataListColumnGenerator } from 'types'
import { StateLabel } from 'components'
import { AccountOverviewLink } from 'components/pages/accounts'

const generateRows: DataListRowGenerator = (accounts: IDeveloperAccount[]) => {
  // Rows and Columns must have the same order
  const mapAccountToRowCell = (account: IDeveloperAccount) => [
    {
      stringValue: account.orgName,
      title: <AccountOverviewLink account={account} />
    },
    account.adminName,
    account.createdAt,
    {
      stringValue: account.state,
      title: <StateLabel state={account.state} />
    }
  ]

  return accounts.map((a) => ({
    id: a.id,
    cells: mapAccountToRowCell(a),
    selected: false
  }))
}

// Filterable columns must have an id equal to its category name
const generateColumns: DataListColumnGenerator = (t) => [
  {
    categoryName: 'group',
    title: t('accountsIndex:accounts_table.group_header'),
    transforms: [sortable, cellWidth(25)]
  },
  {
    categoryName: 'admin',
    title: t('accountsIndex:accounts_table.admin_header'),
    transforms: [sortable, cellWidth(30)]
  },
  {
    categoryName: 'signup',
    title: t('accountsIndex:accounts_table.signup_header'),
    transforms: [sortable, cellWidth(15)]
  },
  {
    categoryName: 'state',
    title: t('accountsIndex:accounts_table.state_header'),
    transforms: [sortable, cellWidth(15)]
  }
]

export { generateColumns, generateRows }
