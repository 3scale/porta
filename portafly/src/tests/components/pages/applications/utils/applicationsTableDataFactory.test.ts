import { generateColumns, generateRows } from 'components/pages/applications/utils'
import { factories } from 'tests/factories'

it('should generate columns', () => {
  const columns = generateColumns(jest.fn())
  expect(columns).toMatchSnapshot()
})

it('should generate rows', () => {
  const applications = factories.Application.buildList(1)
  const columns = generateRows(applications)
  expect(columns).toMatchSnapshot()
})
