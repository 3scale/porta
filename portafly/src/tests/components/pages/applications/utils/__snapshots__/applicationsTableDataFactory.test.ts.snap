// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate columns 1`] = `
Array [
  Object {
    "categoryName": "name",
    "title": undefined,
    "transforms": Array [
      [Function],
      [Function],
    ],
  },
  Object {
    "categoryName": "account",
    "title": undefined,
    "transforms": Array [
      [Function],
      [Function],
    ],
  },
  Object {
    "categoryName": "plan",
    "title": undefined,
    "transforms": Array [
      [Function],
      [Function],
    ],
  },
  Object {
    "categoryName": "created_on",
    "title": undefined,
    "transforms": Array [
      [Function],
      [Function],
    ],
  },
  Object {
    "categoryName": "state",
    "title": undefined,
    "transforms": Array [
      [Function],
      [Function],
    ],
  },
]
`;

exports[`should generate rows 1`] = `
Array [
  Object {
    "cells": Array [
      Object {
        "stringValue": "Cool App 0",
        "title": <ApplicationPageLink
          application={
            Object {
              "account": Object {
                "id": 0,
                "org_name": "Ramdon Org 0",
              },
              "created_on": "2020-06-01T00:00:00Z",
              "id": 0,
              "name": "Cool App 0",
              "plan": Object {
                "id": 0,
                "name": "Super Plan 0",
              },
              "state": "live",
              "traffic_on": "2020-06-01T00:00:00Z",
            }
          }
        />,
      },
      Object {
        "stringValue": "Ramdon Org 0",
        "title": <AccountOverviewLink
          account={
            Object {
              "id": 0,
              "org_name": "Ramdon Org 0",
            }
          }
        />,
      },
      Object {
        "stringValue": "Super Plan 0",
        "title": <PlanOverviewLink
          plan={
            Object {
              "id": 0,
              "name": "Super Plan 0",
            }
          }
        />,
      },
      "2020-06-01T00:00:00Z",
      Object {
        "stringValue": "live",
        "title": <StateLabel
          state="live"
        />,
      },
    ],
    "id": 0,
    "selected": false,
  },
]
`;
